File: hedera-mirror-rest/__tests__/controllers/networkController.test.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import * as constants from '../../constants';
import {NetworkController} from '../../controllers';
import {FileData} from '../../model';
import networkCtrl from '../../controllers/networkController';
import * as utils from '../../utils';

describe('extractNetworkNodesQuery', () => {
  const defaultExpected = {
    conditions: [],
    params: [],
    order: constants.orderFilterValues.ASC,
    limit: 10,
  };

  const specs = [
    {
      name: 'limit',
      input: {
        filters: [
          {
            key: constants.filterKeys.LIMIT,
            operator: utils.opsMap.eq,
            value: 20,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        limit: 20,
        params: ['102'],
      },
    },
    {
      name: 'order desc',
      input: {
        filters: [
          {
            key: constants.filterKeys.ORDER,
            operator: utils.opsMap.eq,
            value: constants.orderFilterValues.DESC,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        order: constants.orderFilterValues.DESC,
        conditions: [],
        params: ['102'],
      },
    },
    {
      name: 'node.id',
      input: {
        filters: [
          {
            key: constants.filterKeys.NODE_ID,
            operator: utils.opsMap.eq,
            value: '10',
          },
        ],
      },
      expected: {
        ...defaultExpected,
        conditions: ['abe.node_id in ($2)'],
        params: ['102', '10'],
      },
    },
  ];

  specs.forEach((spec) => {
    test(`${spec.name}`, () => {
      expect(networkCtrl.extractNetworkNodesQuery(spec.input.filters)).toEqual(spec.expected);
    });
  });
});

describe('validateExtractNetworkNodesQuery throw', () => {
  const specs = [
    {
      name: 'file.id ne',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.ne,
            value: '1000',
          },
        ],
      },
    },
    {
      name: 'file.id gt',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.gt,
            value: '1000',
          },
        ],
      },
    },
    {
      name: 'file.id gte',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.gte,
            value: '1000',
          },
        ],
      },
    },
    {
      name: 'file.id lte',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.lt,
            value: '1000',
          },
        ],
      },
    },
    {
      name: 'file.id lte',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.lte,
            value: '1000',
          },
        ],
      },
    },
    {
      name: 'multi file.id eq',
      input: {
        filters: [
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.eq,
            value: '101',
          },
          {
            key: constants.filterKeys.FILE_ID,
            operator: utils.opsMap.eq,
            value: '102',
          },
        ],
      },
    },
  ];

  specs.forEach((spec) => {
    test(`${spec.name}`, () => {
      expect(() => NetworkController.extractNetworkNodesQuery(spec.input.filters)).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('extractFileDataQuery', () => {
  const defaultExpected = {
    filterQuery: {
      whereQuery: [],
    },
    order: constants.orderFilterValues.ASC,
  };

  const specs = [
    {
      name: 'no timestamp',
      input: {
        filters: [],
      },
      expected: {
        ...defaultExpected,
      },
    },
    {
      name: 'le timestamp',
      input: {
        filters: [
          {
            key: constants.filterKeys.TIMESTAMP,
            operator: utils.opsMap.lt,
            value: 2,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        filterQuery: {
          whereQuery: [
            {
              query: `${FileData.CONSENSUS_TIMESTAMP}  < `,
              param: 2,
            },
          ],
        },
      },
    },
    {
      name: 'lte timestamp',
      input: {
        filters: [
          {
            key: constants.filterKeys.TIMESTAMP,
            operator: utils.opsMap.lte,
            value: 2,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        filterQuery: {
          whereQuery: [
            {
              query: `${FileData.CONSENSUS_TIMESTAMP}  <= `,
              param: 2,
            },
          ],
        },
      },
    },
    {
      name: 'gt timestamp',
      input: {
        filters: [
          {
            key: constants.filterKeys.TIMESTAMP,
            operator: utils.opsMap.gt,
            value: 2,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        filterQuery: {
          whereQuery: [
            {
              query: `${FileData.CONSENSUS_TIMESTAMP}  > `,
              param: 2,
            },
          ],
        },
      },
    },
    {
      name: 'gte timestamp',
      input: {
        filters: [
          {
            key: constants.filterKeys.TIMESTAMP,
            operator: utils.opsMap.gte,
            value: 2,
          },
        ],
      },
      expected: {
        ...defaultExpected,
        filterQuery: {
          whereQuery: [
            {
              query: `${FileData.CONSENSUS_TIMESTAMP}  >= `,
              param: 2,
            },
          ],
        },
      },
    },
  ];

  specs.forEach((spec) => {
    test(`${spec.name}`, () => {
      expect(networkCtrl.extractFileDataQuery(spec.input.filters)).toEqual(spec.expected);
    });
  });
});

describe('convertToCurrencyFormat', () => {
  it.each`
    tinycoins                | currencyFormat | expected
    ${'1234567890000'}       | ${'BOTH'}      | ${'12345.67890000'}
    ${'1234567890000'}       | ${null}        | ${'12345.67890000'}
    ${'0'}                   | ${'BOTH'}      | ${'0.00000000'}
    ${'42'}                  | ${'BOTH'}      | ${'0.00000042'}
    ${'987654321098765432'}  | ${'BOTH'}      | ${'9876543210.98765432'}
    ${'5000000000000000000'} | ${null}        | ${'50000000000.00000000'}
    ${'1234567890000'}       | ${'HBARS'}     | ${'12345'}
    ${'0'}                   | ${'HBARS'}     | ${'0'}
    ${'42'}                  | ${'HBARS'}     | ${'0'}
    ${'987654321098765432'}  | ${'HBARS'}     | ${'9876543210'}
    ${'5000000000000000000'} | ${'HBARS'}     | ${'50000000000'}
    ${'1234567890123'}       | ${'TINYBARS'}  | ${'1234567890123'}
    ${'0'}                   | ${'TINYBARS'}  | ${'0'}
    ${'42'}                  | ${'TINYBARS'}  | ${'42'}
    ${'987654321098765432'}  | ${'TINYBARS'}  | ${'987654321098765432'}
    ${'5000000000000000000'} | ${'TINYBARS'}  | ${'5000000000000000000'}
    ${''}                    | ${undefined}   | ${'0.00000000'}
    ${undefined}             | ${undefined}   | ${'0.00000000'}
    ${undefined}             | ${'HBARS'}     | ${'0'}
    ${undefined}             | ${'TINYBARS'}  | ${'0'}
  `('verifies "$currencyFormat" on $tinycoins expecting $expected', ({tinycoins, currencyFormat, expected}) => {
    expect(networkCtrl.convertToCurrencyFormat(tinycoins, currencyFormat)).toEqual(expected);
  });
});

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/controller/NetworkControllerTest.java
/*
 * Copyright (C) 2022-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.grpc.controller;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.ByteString;
import com.hedera.mirror.api.proto.AddressBookQuery;
import com.hedera.mirror.api.proto.ReactorNetworkServiceGrpc;
import com.hedera.mirror.common.domain.DomainBuilder;
import com.hedera.mirror.common.domain.addressbook.AddressBook;
import com.hedera.mirror.common.domain.addressbook.AddressBookEntry;
import com.hedera.mirror.common.domain.addressbook.AddressBookServiceEndpoint;
import com.hedera.mirror.common.domain.entity.EntityId;
import com.hedera.mirror.grpc.GrpcIntegrationTest;
import com.hedera.mirror.grpc.util.ProtoUtil;
import com.hederahashgraph.api.proto.java.FileID;
import com.hederahashgraph.api.proto.java.NodeAddress;
import com.hederahashgraph.api.proto.java.ServiceEndpoint;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import jakarta.annotation.Resource;
import java.net.InetAddress;
import java.time.Duration;
import java.util.HashSet;
import lombok.CustomLog;
import net.devh.boot.grpc.client.inject.GrpcClient;
import org.apache.commons.lang3.StringUtils;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

@CustomLog
class NetworkControllerTest extends GrpcIntegrationTest {

    private static final Duration WAIT = Duration.ofSeconds(10L);
    private static final long CONSENSUS_TIMESTAMP = 1L;

    @GrpcClient("local")
    private ReactorNetworkServiceGrpc.ReactorNetworkServiceStub reactiveService;

    @Resource
    private DomainBuilder domainBuilder;

    @Test
    void missingFileId() {
        AddressBookQuery query = AddressBookQuery.newBuilder().build();
        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .expectErrorSatisfies(t -> assertException(t, Status.Code.INVALID_ARGUMENT, "fileId: must not be null"))
                .verify(WAIT);
    }

    @Test
    void invalidFileId() {
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder().setFileNum(-1).build())
                .build();
        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .expectErrorSatisfies(t -> assertException(t, Status.Code.INVALID_ARGUMENT, "Invalid entity ID"))
                .verify(WAIT);
    }

    @Test
    void invalidLimit() {
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder().build())
                .setLimit(-1)
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .expectErrorSatisfies(t -> assertException(
                        t, Status.Code.INVALID_ARGUMENT, "limit: must be greater " + "than or equal to 0"))
                .verify(WAIT);
    }

    @Test
    void notFound() {
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder().setFileNum(102L).build())
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .expectErrorSatisfies(t -> assertException(t, Status.Code.NOT_FOUND, "does not exist"))
                .verify(WAIT);
    }

    @Test
    void noLimit() {
        AddressBook addressBook = addressBook();
        AddressBookEntry addressBookEntry1 = addressBookEntry();
        AddressBookEntry addressBookEntry2 = addressBookEntry();
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder()
                        .setFileNum(addressBook.getFileId().getNum())
                        .build())
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .consumeNextWith(n -> assertEntry(addressBookEntry1, n))
                .consumeNextWith(n -> assertEntry(addressBookEntry2, n))
                .expectComplete()
                .verify(WAIT);
    }

    @Test
    void noLimitServiceEndpointWithDomainName() {
        var addressBook = addressBook();
        var addressBookEntry1 = addressBookEntryCustomized("www.example-node.com", "", 5000);

        var query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder()
                        .setFileNum(addressBook.getFileId().getNum())
                        .build())
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .consumeNextWith(n -> assertEntry(addressBookEntry1, n))
                .expectComplete()
                .verify(WAIT);
    }

    @Test
    void testWithEmptyDomainNameAndIpAddress() {
        var addressBook = addressBook();
        var addressBookEntry1 = addressBookEntryCustomized("", "", 0);

        var query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder()
                        .setFileNum(addressBook.getFileId().getNum())
                        .build())
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .consumeNextWith(n -> assertEntry(addressBookEntry1, n))
                .expectComplete()
                .verify(WAIT);
    }

    @Test
    void limitReached() {
        AddressBook addressBook = addressBook();
        AddressBookEntry addressBookEntry1 = addressBookEntry();
        addressBookEntry();
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder()
                        .setFileNum(addressBook.getFileId().getNum())
                        .build())
                .setLimit(1)
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .consumeNextWith(n -> assertEntry(addressBookEntry1, n))
                .expectComplete()
                .verify(WAIT);
    }

    @SuppressWarnings("deprecation")
    @Test
    void nullFields() {
        AddressBook addressBook = addressBook();
        AddressBookEntry addressBookEntry = domainBuilder
                .addressBookEntry()
                .customize(a -> a.consensusTimestamp(CONSENSUS_TIMESTAMP)
                        .description(null)
                        .memo(null)
                        .nodeCertHash(null)
                        .publicKey(null)
                        .stake(null))
                .persist();
        AddressBookQuery query = AddressBookQuery.newBuilder()
                .setFileId(FileID.newBuilder()
                        .setFileNum(addressBook.getFileId().getNum())
                        .build())
                .build();

        StepVerifier.withVirtualTime(() -> reactiveService.getNodes(Mono.just(query)))
                .thenAwait(WAIT)
                .consumeNextWith(n -> assertThat(n)
                        .isNotNull()
                        .returns("", NodeAddress::getDescription)
                        .returns(ByteString.EMPTY, NodeAddress::getMemo)
                        .returns(addressBookEntry.getNodeAccountId(), t -> EntityId.of(n.getNodeAccountId()))
                        .returns(ByteString.EMPTY, NodeAddress::getNodeCertHash)
                        .returns(addressBookEntry.getNodeId(), NodeAddress::getNodeId)
                        .returns("", NodeAddress::getRSAPubKey)
                        .returns(0L, NodeAddress::getStake))
                .expectComplete()
                .verify(WAIT);
    }

    private AddressBook addressBook() {
        return domainBuilder
                .addressBook()
                .customize(a -> a.startConsensusTimestamp(CONSENSUS_TIMESTAMP))
                .persist();
    }

    private AddressBookEntry addressBookEntry() {
        return domainBuilder
                .addressBookEntry(1)
                .customize(a -> a.consensusTimestamp(CONSENSUS_TIMESTAMP))
                .persist();
    }

    private AddressBookEntry addressBookEntryCustomized(String domainName, String ipAddress, int port) {
        var serviceEndpoints = new HashSet<AddressBookServiceEndpoint>();
        var endpoint = domainBuilder
                .addressBookServiceEndpoint()
                .customize(a -> a.domainName(domainName).ipAddressV4(ipAddress).port(port))
                .get();
        serviceEndpoints.add(endpoint);
        return domainBuilder
                .addressBookEntry(1)
                .customize(a -> a.serviceEndpoints(serviceEndpoints).consensusTimestamp(CONSENSUS_TIMESTAMP))
                .persist();
    }

    @SuppressWarnings("deprecation")
    private void assertEntry(AddressBookEntry addressBookEntry, NodeAddress nodeAddress) {
        assertThat(nodeAddress)
                .isNotNull()
                .returns(addressBookEntry.getDescription(), NodeAddress::getDescription)
                .returns(ByteString.copyFromUtf8(addressBookEntry.getMemo()), NodeAddress::getMemo)
                .returns(addressBookEntry.getNodeAccountId(), n -> EntityId.of(n.getNodeAccountId()))
                .returns(ProtoUtil.toByteString(addressBookEntry.getNodeCertHash()), NodeAddress::getNodeCertHash)
                .returns(addressBookEntry.getNodeId(), NodeAddress::getNodeId)
                .returns(addressBookEntry.getPublicKey(), NodeAddress::getRSAPubKey)
                .returns(addressBookEntry.getStake(), NodeAddress::getStake);

        var serviceEndpoint = addressBookEntry.getServiceEndpoints().iterator().next();
        ByteString ipAddress = ByteString.EMPTY;
        try {
            if (StringUtils.isNotBlank(serviceEndpoint.getIpAddressV4())) {
                ipAddress = ProtoUtil.toByteString(
                        InetAddress.getByName(serviceEndpoint.getIpAddressV4()).getAddress());
            }
        } catch (Exception e) {
            // Ignore
        }
        assertThat(nodeAddress.getServiceEndpointList())
                .hasSize(1)
                .first()
                .returns(ipAddress, ServiceEndpoint::getIpAddressV4)
                .returns(serviceEndpoint.getPort(), ServiceEndpoint::getPort)
                .returns(serviceEndpoint.getDomainName(), ServiceEndpoint::getDomainName)
                .extracting(ServiceEndpoint::getIpAddressV4)
                .isNotEqualTo(
                        ByteString.copyFrom(InetAddress.getLoopbackAddress().getAddress()));
    }

    private void assertException(Throwable t, Status.Code status, String message) {
        assertThat(t).isNotNull().isInstanceOf(StatusRuntimeException.class).hasMessageContaining(message);

        StatusRuntimeException statusRuntimeException = (StatusRuntimeException) t;
        assertThat(statusRuntimeException.getStatus().getCode()).isEqualTo(status);
    }
}
